TuningApp.py

```diff
@@ -0,0 +1,592 @@
+from ctt_macbeth_locator import find_macbeth, get_macbeth_chart
+
+
+# doesnt let runtime warning print
+def fxn():
+    warnings.warn("runtime", RuntimeWarning)
+
+
+# Configuring the camera
+# sorting functions for camera modes
+def sortFn(d):
+    return d['crop_limits'][-1] * d['crop_limits'][-2]
+
+
+def sortBitDepth(d):
+    return d['bit_depth']
+
+
+picam2 = Picamera2()
+candidate_modes = []
+
+
+for x in picam2.sensor_modes:
+    if 0 in x['crop_limits']:
+        candidate_modes.append(x)
+
+
+if len(candidate_modes) == 1:
+    mode = candidate_modes[0]
+else:
+    candidate_modes.sort(key=sortFn)
+    tuple_ful = candidate_modes[-1]['size']
+    tuple_half = tuple(ti / 2 for ti in tuple_ful)
+
+    half_candidates = [mode for mode in candidate_modes if x['size'] == tuple_half]
+    full_candidates = [mode for mode in candidate_modes if x['size'] == tuple_ful]
+
+    if half_candidates:
+        half_candidates.sort(key=sortBitDepth)
+        mode = half_candidates[-1]
+    else:
+        full_candidates.sort(key=sortBitDepth)
+        mode = full_candidates[-1]
+main = {"size": (800, 600)}
+sensor = {'output_size': mode['size'], 'bit_depth': mode['bit_depth']}
+config = picam2.create_preview_configuration(main=main, sensor=sensor)
+picam2.configure(config)
+
+
+# Function that returns true if it can detect macbeth plot
+def check(filename, target):
+    Cam = Camera("imx.json", json=json_template)
+    Img = dng_load_image(Cam, filename)
+
+    av_chan = (np.mean(np.array(Img.channels), axis=0) / (2 ** 16))
+    av_val = np.mean(av_chan)
+
+    if av_val < Img.blacklevel_16 / (2**16) + 1 / 64:
+        return False
+    else:
+        macbeth = find_macbeth(Cam, av_chan, mac_config=(0, 0))
+    if macbeth is None:
+        return False
+    else:
+        return True
+
+
+# add camera name to this folder that is default
+index_photo = 0
+start = False
+list_of_index = []
+for f in os.listdir(os.path.expanduser("~")):
+    if picam2.global_camera_info()[0]["Model"] in f:
+        if "_" in f:
+            list_of_index.append(int(f[f.rindex("_") + 1:]))
+        index_photo += 1
+if index_photo == 0:
+    folder_directory = os.path.join(os.path.expanduser("~"), picam2.global_camera_info()[0]["Model"])
+elif list_of_index == []:
+    folder_directory = os.path.join(os.path.expanduser("~"), picam2.global_camera_info()[0]["Model"] + "_" + str(index_photo))
+else:
+    list_of_index.sort()
+    index_photo = list_of_index[-1] + 1
+    folder_directory = os.path.join(os.path.expanduser("~"), picam2.global_camera_info()[0]["Model"] + "_" + str(index_photo))
+
+# what pi are we working on?
+if Picamera2.platform == Platform.PISP:
+    target = "pi5"
+    from ctt_pisp import json_template
+else:
+    target = "pi4"
+    from ctt_vc4 import json_template
+
+
+# checks if sting is a valid path to a folder
+def valid_path(s):
+    if (s[0] != "/" or s[-1] == "/"):
+        return False
+    else:
+        return True
+
```

Comment:
I guess I wonder a little bit whether there's a "standard" Python function that does this? Might be worth a quick web search to check, but otherwise I'm not too bothered about this for the moment!

Reply:
Tried finding it last week, but could not